ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/operations/SLoadOperation.java

```diff
@@ -15,27 +15,60 @@
 package org.hyperledger.besu.ethereum.vm.operations;
 
 import org.hyperledger.besu.ethereum.core.Account;
+import org.hyperledger.besu.ethereum.core.Address;
+import org.hyperledger.besu.ethereum.core.Gas;
+import org.hyperledger.besu.ethereum.vm.AbstractOperation;
 import org.hyperledger.besu.ethereum.vm.EVM;
+import org.hyperledger.besu.ethereum.vm.ExceptionalHaltReason;
+import org.hyperledger.besu.ethereum.vm.FixedStack.OverflowException;
+import org.hyperledger.besu.ethereum.vm.FixedStack.UnderflowException;
 import org.hyperledger.besu.ethereum.vm.GasCalculator;
 import org.hyperledger.besu.ethereum.vm.MessageFrame;
 
+import java.util.Optional;
+
 import org.apache.tuweni.bytes.Bytes32;
 import org.apache.tuweni.units.bigints.UInt256;
 
-public class SLoadOperation extends AbstractFixedCostOperation {
+public class SLoadOperation extends AbstractOperation {
+
+  private final Optional<Gas> warmCost;
+  private final Optional<Gas> coldCost;
+
+  private final OperationResult warmSuccess;
+  private final OperationResult coldSuccess;
 
   public SLoadOperation(final GasCalculator gasCalculator) {
-    super(0x54, "SLOAD", 1, 1, false, 1, gasCalculator, gasCalculator.getSloadOperationGasCost());
+    super(0x54, "SLOAD", 1, 1, false, 1, gasCalculator);
+    final Gas baseCost = gasCalculator.getSloadOperationGasCost();
+    warmCost = Optional.of(baseCost.plus(gasCalculator.getWarmStorageReadCost()));
+    coldCost = Optional.of(baseCost.plus(gasCalculator.getColdSloadCost()));
+
+    warmSuccess = new OperationResult(warmCost, Optional.empty());
+    coldSuccess = new OperationResult(coldCost, Optional.empty());
   }
 
   @Override
-  public OperationResult executeFixedCostOperation(final MessageFrame frame, final EVM evm) {
-    final Bytes32 key = frame.popStackItem();
-
-    final Account account = frame.getWorldState().get(frame.getRecipientAddress());
-
-    frame.pushStackItem(account.getStorageValue(UInt256.fromBytes(key)).toBytes());
+  public OperationResult execute(final MessageFrame frame, final EVM evm) {
+    try {
+      final Account account = frame.getWorldState().get(frame.getRecipientAddress());
+      final Address address = account.getAddress();
+      final Bytes32 key = frame.popStackItem();
+      final boolean slotIsWarm = frame.warmUpStorage(address, key);
+      final Optional<Gas> optionalCost = slotIsWarm ? warmCost : coldCost;
```

Comment:
This might be related to a question I asked in a previous pr but I can't remember. Do these need to be optional everywhere? For example, isn't the warmCost and coldCost always present here?

Reply:
It's a mix.  We need the optional for the return value but we need the value during gas calculations.  The gas cost is not optional within the method but on the return, for some errors we return an empty cost (such as when we can't know because of a stack underflow).

Reply:
Ah, I see, could we push the optional wrapping to as late as possible? I think it makes it clearer for which sections it is guaranteed to be defined. The intent is a bit unclear when the warmCost and coldCost fields are Optional.

Reply:
The optionals are re-used across requests.  If we wrapped in the execute method it would result in object churn.